DeviceCommunications/DeviceComsProtocolParser.cs

```diff
@@ -0,0 +1,194 @@
+        IDeviceComsStream _DeviceComsStream;
+
+        private object _dataPointsLock = new object();
+
+        private readonly ConcurrentQueue<(long, EEGData)> _dataPoints;
+
+        public DeviceComsProtocolParser(IDeviceComsStream deviceComsStream)
+        {
+            _DeviceComsStream = deviceComsStream;
+
+            if (_DeviceComsStream.Connect())
+            {
+                // if device is connected then subscribe to it's DataReceived events
+                _DeviceComsStream.DataReceivedEvent += DataReceivedEventHandler;
+            }
+
+            _dataPoints = new ConcurrentQueue<(long, EEGData)>();
+        }
+
+        public event EventHandler DataUpdated;
+
+        void IDeviceComs.SendGetDataSinceLastRequest()
+        {
+            byte[] data = new byte[3];
+
+            data[0] = 0xFF;
+            data[1] = 0xFF;
+            data[2] = 0x03;
+
+            _DeviceComsStream.TransmitData(data);
+        }
+
+        void IDeviceComs.SendStartDataCaptureRequest()
+        {
+            byte[] data = new byte[3];
+
+            data[0] = 0xFF;
+            data[1] = 0xFF;
+            data[2] = 0x01;
+
+            _DeviceComsStream.TransmitData(data);
+        }
+
+        void IDeviceComs.SendStopDataCaptureRequest()
+        {
+            byte[] data = new byte[3];
+
+            data[0] = 0xFF;
+            data[1] = 0xFF;
+            data[2] = 0x02;
+
+            _DeviceComsStream.TransmitData(data);
+        }
+
+        private int _RxIndex = 0;
+        private byte[] _RxData = new byte[27];
+
+        EEGSignalMetadata IEEGData.SignalMetadata => throw new NotImplementedException();
+
+        long _LastUpdateTime;
+        long IEEGData.LastUpdateTime { get { return _LastUpdateTime; } }
+
+        /// <summary>
+        /// Also determines the interval at which this MockEEGData class will 
+        /// generate data.
+        /// Changing this has no effect if the timer is running.
+        /// </summary>
+        double IEEGData.DigitalMinimum { get; } = -1;
+
+        double IEEGData.DigitalMaximum { get; } = 1;
+
+        double IEEGData.PhysicalMinimum { get; } = -20;
+
+        double IEEGData.PhysicalMaximum { get; } = 20;
+
+        EEGSignalMetadata SignalMetadata { get; } = new EEGSignalMetadata()
+        {
+            Label = "EEG",
+            Units = "mV",
+            TransducerType = "UNKNOWN",
+            Prefiltering = "UNKNOWN",
+        };
+
+        // Returns a copy whenever it's accessed, to avoid concurrency issues.
+        public IEnumerable<(long, EEGData)> DataPoints
+        {
+            get
+            {
+                lock (_dataPointsLock)
+                {
+                    return _dataPoints.ToArray();
+                }
+            }
+        }
+
+        private void DataReceivedEventHandler(object sender, EventArgs e)
+        {
+            byte[] rxData;
+
+            int count = _DeviceComsStream.GetReceivedData(out rxData);
```

Comment:
What would you think about having the data received be sent through some custom EventArgs instead?
(not a change that needs to be made in this PR, but something I'd like to do and want feedback on, which will be the case for basically all of these comments)

Reply:
What are the advantages to this? I don't know enough to comment really, this does seem like a very firmwarey way of doing it.

Reply:
It pretty much just avoids having to call another function to retrieve the data.
That does make sense to me though, as in firmware you would have to allocate an array for the 'get data' method to fill with its data?